compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3553,6 +3555,70 @@ TR::Register *OMR::Power::TreeEvaluator::vmulInt32Helper(TR::Node *node, TR::Cod
    return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmuluwm);
    }
 
+TR::Register *OMR::Power::TreeEvaluator::vmulInt64Helper(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P10))
+      return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmulld);
+
+   /*
+   To perform the multiplication without the vmulld instruction, we will break it into parts.
+
+   Suppose we have two long integers (64 bits), AB and CD, where each of A, B, C, and D is one word (32 bits).
+   To get their product, AB * CD, we can do the following:
+
+   AB * CD = B*D + 2^32*(AD + BC) + 2^64*(A*C)
+
+   Since we are only keeping the lower 64 bits of the result, this is equivalent to:
+
+   AB * CD = B*D + 2^32*(AD + BC)
+   */
+
+   TR::Node *firstChild;
+   TR::Node *secondChild;
+   TR::Register *lhsReg, *rhsReg;
+   TR::Register *productReg;
+   TR::Register *tempA, *tempB;
+   TR::Register *shiftReg;
+
+   firstChild = node->getFirstChild();
+   secondChild = node->getSecondChild();
+
+   lhsReg = cg->evaluate(firstChild);
+   rhsReg = cg->evaluate(secondChild);
+
+   productReg = cg->allocateRegister(TR_VRF);
+   tempA = cg->allocateRegister(TR_VRF);
+   tempB = cg->allocateRegister(TR_VRF);
+   shiftReg = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(productReg);
+
+   //set shift amount to 16 bits
```

Comment:
This is a very interesting idea but needs to be documented in the comment. I think "set shift amount to 16 bits" is misleading.

Reply:
I think this is a holdover from when I achieved a 32 bit shift by shifting by 16 bits twice, rather than using Jimmy's suggestion. Either way though, you're right that it's misleading and should be corrected